### PR TITLE
[WIP] Implement `scipy.spatial.distance`

### PR DESCRIPTION
**Pull request by martinkim0**
Partially addresses #9235. I saw that there's a similar PR #16147, but it looks like it's inactive.

To-do:

- [x] Metric functions
- [ ] `scipy.spatial.distance.cdist`
- [ ] `scipy.spatial.distance.pdist`
- [ ] `scipy.spatial.distance.squareform`
- [ ] `scipy.spatial.distance.directed_hausdorff`
- [ ] Utility functions (`is_valid_dm`, `is_valid_y`, `num_obs_dm`, `num_obs_y`)
- [ ] Clean up metric functions
- [ ] `rogerstanimono` and `sokalmichener` are the same? https://github.com/scipy/scipy/issues/2011
- [ ] Add tests
- [ ] Add docs

### CODE REVIEW

**Comment by jakevdp** [comment]
Hi - thanks for working on this! Just a note: it looks like the implementations are copy-pasted from the scipy source code. This is problematic from a code licensing standpoint – for this sort of work I'd recommend not referencing the scipy source, but rather implementing the metrics from scratch, using unit tests to ensure that the implementations produce similar results.

**Comment by martinkim0** [comment]
@jakevdp Hi, thanks for the feedback! Makes sense, I'll take a second pass at this and implement the metrics from scratch.

**Comment by jakevdp** [comment]
Hi again: we've done some thinking about the scope of `jax.scipy` long-term, and the consensus was that `scipy.spatial` is out of scope. See https://jax.readthedocs.io/en/latest/jep/18137-numpy-scipy-scope.html for details.

Thanks so much for the contribution, and I'm sorry we weren't able to merge your code!
